### PR TITLE
GG-33352 Add resource de-allocation in SharedPageLockTracker

### PR DESCRIPTION
**Pull request by sashapolo**
https://ggsystems.atlassian.net/browse/GG-33352

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IndexStorageImpl.java [hunk 1]
```diff
@@ -251,26 +263,26 @@ private String maskCacheIndexName(Integer cacheId, String idxName, int segment)
...
+            IOVersions<? extends BPlusInnerIO<IndexItem>> innerIos,
+            IOVersions<? extends BPlusLeafIO<IndexItem>> leafIos,
+            boolean initNew,
             @Nullable FailureProcessor failureProcessor,
-            @Nullable PageLockListener lockLsnr
+            PageLockTrackerManager pageLockTrackerManager
         ) throws IgniteCheckedException {
             super(
-                treeName("meta", "Meta"),
+                treeName,
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IndexStorageImpl.java:285]: what was this for?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/defragmentation/CachePartitionDefragmentationManager.java [hunk 2]
```diff
@@ -593,7 +590,13 @@ private boolean defragmentOnePartition(
 
     /** */
     public IgniteInternalFuture<?> completionFuture() {
-        return completionFut.chain(future -> null);
+        return completionFut.chain(future -> {
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/defragmentation/CachePartitionDefragmentationManager.java:593]: What if no one's calling this method? I'd move chaining into future creation instead

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/defragmentation/LinkMap.java [hunk 3]
```diff
@@ -45,12 +41,12 @@
     public static final int META_PAGE_IDX = 2;
 
     /** */
-    public static final IOVersions<? extends BPlusLeafIO<?>> LEAF_IO_VERSIONS = new IOVersions<>(
+    public static final IOVersions<LinkMappingLeafIO> LEAF_IO_VERSIONS = new IOVersions<>(
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/defragmentation/LinkMap.java:44]: This was deliberate, IO classes are private. If you have a better option then go ahead

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/SharedPageLockTracker.java [hunk 4]
```diff
@@ -138,50 +137,46 @@ private PageLockTracker createTracker() {
...
+            // Size for the new (K,V) pair.
+            memCalc.onHeapAllocated((name.getBytes().length + 16) + (8 + 16 + 4));
 
-        if (id == null) {
-            structureNameToId.put(structureName, id = (++idGen));
-
-            // Size for new (K,V) pair.
-            memCalc.onHeapAllocated((structureName.getBytes().length + 16) + (8 + 16 + 4));
-        }
+            return ++idGen;
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/SharedPageLockTracker.java:145]: Not threadsafe

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToFileDumpProcessor.java [hunk 5]
```diff
@@ -16,25 +16,26 @@
...
-import static java.nio.file.StandardOpenOption.WRITE;
+import static org.apache.ignite.internal.processors.cache.persistence.diagnostic.pagelocktracker.dumpprocessors.ToStringDumpHelper.toStringDump;
 
 /**
  *
  */
 public class ToFileDumpProcessor {
     /** Date format. */
-    public static final SimpleDateFormat DATE_FMT = new SimpleDateFormat("yyyy_MM_dd_HH_mm_ss_SSS");
+    public static final DateTimeFormatter DATE_FMT = DateTimeFormatter
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToFileDumpProcessor.java:36]: Is this one finally threadsafe? Cause it feels like we had a bug here

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpHelper.java [hunk 6]
```diff
@@ -24,7 +24,6 @@
  * Helper for creating string from {@link PageLockDump}.
  */
 public class ToStringDumpHelper {
-
     /** Date format. */
     public static final SimpleDateFormat DATE_FMT = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS");
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpHelper.java:28]: I know it's out of scope, but can you please also check usages of this formatter?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/freelist/CacheFreeList.java [hunk 7]
```diff
@@ -75,9 +76,4 @@ public CacheFreeList(
             "Constructed a link with invalid partition ID [partId=" + row.key().partition() +
                 ", link=" + U.hexLong(row.link()) + ']';
     }
-
-    /** {@inheritDoc} */
-    @Override public String toString() {
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/freelist/CacheFreeList.java:80]: Interesting, why did you remove this method?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToFileDumpProcessor.java (hunk 5 above)
Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToFileDumpProcessor.java:36]: yes, that's why I changed it

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/freelist/CacheFreeList.java (hunk 7 above)
Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/freelist/CacheFreeList.java:80]: it exactly duplicates the method from its superclass

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IndexStorageImpl.java (hunk 1 above)
Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IndexStorageImpl.java:285]: what was what for?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/SharedPageLockTracker.java (hunk 4 above)
Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/SharedPageLockTracker.java:145]: wow, nice catch!

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpHelper.java (hunk 6 above)
Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpHelper.java:28]: I've fixed this particular usage, but let me change other usages in a separate PR

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IndexStorageImpl.java (hunk 1 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IndexStorageImpl.java:285]: Sorry, are there any clues about that original line? What was its purpose?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpHelper.java (hunk 6 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpHelper.java]: Sure

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/freelist/CacheFreeList.java (hunk 7 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/freelist/CacheFreeList.java:80]: Ok

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IndexStorageImpl.java (hunk 1 above)
Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IndexStorageImpl.java:285]: it was a "name for debugging purposes", e.g. it was printed in `toString()`. A name will still be printed, it will just be a little different (but very similar)

**Review by ibessonov — APPROVED** [review]
